packages/edit-site/src/components/sidebar/global-styles-v2/state.js

```diff
@@ -0,0 +1,183 @@
+/**
+ * External dependencies
+ */
+import { v4 as uuid } from 'uuid';
+import { get, kebabCase, set, startCase } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	createContext,
+	useCallback,
+	useContext,
+	useState,
+} from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { COLOR_ACCENT, COLOR_MAIN, COLOR_TEXT } from './constants';
+
+// Models
+const createPaletteColor = ( { color, slug } ) => ( {
+	id: uuid(),
+	slug,
+	color,
+	title: startCase( slug ),
+} );
+
+const createPalette = ( { colors, title } ) => ( {
+	id: uuid(),
+	title,
+	colors,
+} );
+
+const createElement = ( { title, ...rest } ) => ( {
+	...rest,
+	id: uuid(),
+	title,
+	slug: kebabCase( title ),
+} );
+
+// App State
+const initialState = {
```

Comment:
The state needs a bunch of love. We'll likely need to work on a new `@wordpress/data` store for this one.

Reply:
Will we need to just adopt whatever global styles currently uses for state management? I'm assuming there's a lot of intermediary stuff that needs to be done when the state changes that's already being handled in the current implementation?

Reply:
Yes, it makes sense to reuse as much as possible from the existing global styles UI as a whole, not only the state but also the existing components. Which is one more thing we'll have to consider as we move forward to this version.

Reply:
Yeah we need to see how global styles are read/written and think about the best way to integrate that into the sidebar

Reply:
cc @nosolosw and @jorgefilipecosta here for input

Reply:
A good starting point would be to look at the `useGlobalStylesContext` hook, it gives access to the current settings & styles (see [in use](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/sidebar/global-styles-sidebar.js#L153)). They follow the `theme.json` [specification](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/).

Reply:
Another good place to check the logic is the implementation of the GlobalStylesProvider /packages/edit-site/src/components/editor/global-styles-provider.js.
Basically, the source of truth for the global styles user data is in wordpress/data in a post type that contains global styles information. (retrieved using useGlobalStylesEntityContent that just returns the content of the post type useEntityProp( 'postType', 'wp_global_styles', 'content' );).

The provider contains a JSON object representation of that same data (memoized each time the data changes). And also contains mergedStyles the result of merging the user theme.json with the theme and core theme.json data provided by the server.

The provider provides setters and getters for user styles and settings that deal with variable retrieving etc. The consumer of the API just reads and writes values if the values can be represented with CSS variables or come from CSS variables the consumer is not aware.

On each change, there are also effects on the provider that update the compiled CSS of the theme.json tree so the user sees style updates in real-time.